fastify-swagger static mode

PavelPolyakov opened the pull request:
hi @delvedor,

As we've discussed here is the update for the `fastify-swagger` which allows to serve the static swagger specification.

Please review the overall form, after your comments and if everything is fine I'd update the tests/readme/module version etc.

PavelPolyakov commented:
Hello @delvedor, friendly reminder, could you please share your feedback?

Regards,

delvedor commented:
Hi!
I'm stuck at work and I can't review this right now, I'll do this asap, sorry for the delay.
In the meantime can you fix the test under Node v4 and remove the `package-lock.json` file?
Thanks.

PavelPolyakov commented:
@delvedor sure, sorry for disturbing. Take your time.

reg v4 - I know what is the issue, I'll fix it along with the further work finalising if you approve the overall concept. 

`package-lock.json` - I thought `npm` recommends to commit it. I'll remove it as well.

PavelPolyakov commented:
I've updated README and removed `package-lock.json`.

Though I think you might want to reconsider this, since [here](https://docs.npmjs.com/files/package-lock.json) they recommend to commit it. Or maybe I miss something.

mcollina commented:
@PavelPolyakov here is why: https://github.com/sindresorhus/ama/issues/479#issuecomment-310661514.

PavelPolyakov commented:
hello @delvedor, is there any chance this will be merge soon? I'd like to use it in the project.

delvedor commented on static.js:
```diff
@@ -0,0 +1,86 @@
+
+    // read
+    const source = fs.readFileSync(
+      `${process.cwd()}/${opts.specification.path}`,
```
This is not cross platform, you should use `path.resolve` instead.
Same above :)

PavelPolyakov replied:
updated, thanks

PavelPolyakov commented on tap-snapshots/test-static.js-TAP.test.js:
```diff
@@ -0,0 +1,32 @@
+/* IMPORTANT
```
I think it's common practice to commit snapshots, so they are used during the assertion

mcollina replied:
Yes, this is needed.

PavelPolyakov commented on index.js:
```diff
@@ -1,291 +1,22 @@
 
-  var end = url.indexOf('/', ++start)
+  // by default the mode is dynamic, as plugin initially was developed
+  opts.mode = opts.mode && ['dynamic', 'static'].indexOf(opts.mode.toLowerCase()) !== -1 ? opts.mode.toLowerCase() : 'dynamic'
```
updated, thanks

delvedor commented:
Can you fix linting rules?

PavelPolyakov commented:
Sure, doing this, "it works on my machine" for some reason:
![image](https://user-images.githubusercontent.com/839290/37277399-d2835614-25e4-11e8-8a56-d75883f7b648.png)

delvedor commented:
In master we are using standard v11 :D

PavelPolyakov commented:
all done :)

mcollina commented on README.md:
```diff
@@ -86,53 +86,79 @@ fastify.ready(err => {
+<a name="mode.static"></a>
 
-If you pass `{ exposeRoute: true }` the plugin will expose the documentation with the following apis:
+* `static` mode should be configured explicitly. In this mode `fastify-swagger` serves given specification, you should craft it yourselfe.
```
I would prefer if the dynamic and static modes are made subheadings `###` instead of `*`.

PavelPolyakov replied:
done, but hierarchically it's `#####`

mcollina commented on README.md:
```diff
@@ -86,53 +86,79 @@ fastify.ready(err => {
+  {
+    mode: 'static',
+    specification: {
+      type: 'file',
```
are there other types?

PavelPolyakov replied:
right now not, but I thought that it will be possible to provide `url` or `file`

mcollina replied:
then let's not have a  `type` at all.

PavelPolyakov replied:
done

mcollina commented on index.js:
```diff
@@ -1,291 +1,24 @@
+      setup.dynamic(fastify, opts, next)
+      break
+    default:
+      throw new Error("unsupported mode, should be one of ['static', 'dynamic']")
```
Can you call `next()` with this error instead?

PavelPolyakov replied:
done, thanks

mcollina commented on static.js:
```diff
@@ -0,0 +1,86 @@
+    assert(
+      ['.yaml', '.json'].indexOf(extName) !== -1,
+      "specification.path extension name is not supported, should be one from ['.yaml', '.json']"
+    )
```
Because of the asynchronous loading of fastify, all those asserts should call `next(err)`.

PavelPolyakov replied:
This is a good catch, do you see (or know) if it is possible to achieve this with `assert`? (without me diving into the fastify core code)
Because in the other module, where I actually test the validation, I was forced to refactor it to:
```
  if (!opts.secret) { return next(new Error('fastify-keycloak: secret should be provided')); }
```

if I use `assert(false, next('error'))`, what happens with the exception which is thrown by `assert`?

![image](https://user-images.githubusercontent.com/839290/37281065-07fcef98-25f0-11e8-9fcf-bced037d5fa8.png)

mcollina replied:
You can wrap your asserts with a try/catch block.

PavelPolyakov replied:
yes, I've also tried this, but after thought that `if` and `next(new Error())` is better.

ok, thanks for your comments and feedback, I'll try to fix them later this day.

PavelPolyakov replied:
refactored to `if` + `return next(new Error())`

mcollina commented on test/static.js:
```diff
@@ -0,0 +1,73 @@
+      }
+    }
+  )
+})
```
Can you please add some tests around these validations?

PavelPolyakov replied:
added additional test

PavelPolyakov commented:
all done v2

PavelPolyakov commented:
@delvedor 
sorry for pursuing you, I can imagine you have things to do :) but I really want to have it integrated to the one of the projects I'm working on currently

delvedor commented:
@PavelPolyakov thank you for all you work and sorry for the delay!!

PavelPolyakov commented:
np, thank you, waiting for the new npm then :)
